### PR TITLE
Revise FontContext font loading

### LINKED ISSUE #643: IOS font not appearing
Hello, Im trying to display vector data from url  which is loading properly like buildings and poi but the labels are not appearing below i have pasted the console log please guide what is the issue, im using xcode 7.2.1. and cmake version 3.4.0

TANGRAM tileManager.cpp:67: add source local
TANGRAM tangram.cpp:99: finish initialize
TANGRAM tangram.cpp:577: setup GL
TANGRAM hardware.cpp:56: Driver supports map buffer: 1
TANGRAM hardware.cpp:57: Driver supports vaos: 1
TANGRAM hardware.cpp:71: Hardware max texture size 4096
TANGRAM hardware.cpp:72: Hardware max combined texture units 8
TANGRAM tangram.cpp:132: resize: 1242 x 2208
TANGRAM tileWorker.cpp:70: Passed new TileBuilder to TileWorker
Failed to read file at path: /var/mobile/Containers/Bundle/Application/40249C13-C398-4624-9D27-D6A170598A5E/tangram.app/fonts/default-400normal.ttf
Failed to read file at path: /var/mobile/Containers/Bundle/Application/40249C13-C398-4624-9D27-D6A170598A5E/tangram.app/fonts/default-400normal.ttf
Failed to read file at path: 
ERROR fontContext.cpp:263: Could not load font file default_400_normal
TANGRAM tileWorker.cpp:70: Passed new TileBuilder to TileWorker
Missing font: error: 1 NotoSans-Regular.ttf
Missing font: error: 1 NotoNaskh-Regular.ttf
Missing font: error: 1 NotoSansHebrew-Regular.ttf
Missing font: error: 1 DroidSansJapanese.ttf
Missing font: error: 1 DroidSansFallback.ttf
Missing font: error: 1 NotoSans-Regular.ttf
Missing font: error: 1 NotoNaskh-Regular.ttf
Missing font: error: 1 NotoSansHebrew-Regular.ttf

### PR DESCRIPTION
**Pull request by karimnaaji**
- Remove platform dependent code
- Let platform layer resolve font path and font binary loading
- Fix #643

### CODE REVIEW

**Comment by haidershahzad** [comment]
alfons::InputSource source;
is not working. see attach image.
screen shot 2016-03-29 at 5 19 28 pm

**Comment by karimnaaji** [comment]
Hi @haidershahzad , we're working with git submodules. You always have to checkout those submodules whenever you change to a new branch. In this case `git submodule update` after `git checkout fix-alfons-font-loading` would fix that issue.

File: core/src/text/fontContext.cpp [hunk 1]
```diff
@@ -5,68 +5,92 @@
...
-#define STEP_SIZE 12
-
-#define SDF_WIDTH 6
+namespace Tangram {
 
-#define MIN_LINE_WIDTH 4
+bool addFace(std::shared_ptr<alfons::Font>& font,
+    alfons::FontManager& manager,
+    const FallbackPath& fallback, float size)
+{
```
Comment (tallytalwar) [inline core/src/text/fontContext.cpp]: nitpick but can you update this to follow our function scope styling. Thanks

Comment (karimnaaji) [inline core/src/text/fontContext.cpp]: I'm pretty unclear on what we're following concerning this though.

Comment (karimnaaji) [inline core/src/text/fontContext.cpp]: Ran clang-format on those.

**Comment by blair1618** [comment]
Any reason not to merge this?

Running clang-format has shown that we may want to revise those formatting options, but functionally this looks good to me.

**Comment by hjanetzek** [comment]
@blair1618  ah, the comment about clang format was regarding this PR. 120 chars gets also clipped in github, but mainly I find it much easier to read less wide columns and it would allow to have two editor windows next to each other without side scrolling here :)

**Comment by blair1618** [comment]
@hjanetzek Weird, 120 columns fit in github on my screen. I wouldn't mind reducing our preferred column limit. I know many projects use 80, but that feels a bit tight with 4-space indent. 100 sounds like a good middle ground.

**Comment by tallytalwar** [comment]
Actually reviewing this now!

**Comment by tallytalwar** [comment]
On the line wrapping!! I prefer 120, but can live with 100 :D.

**Comment by hjanetzek** [comment]
Loading all the fallback fonts for android into memory increases native heap usage from 12 to 66MB here. It would be possible to reduce this by disposing the InputSource once loaded (since we never unload and reload Fonts). Still this is quite a lot

**Comment by hjanetzek** [comment]
```
[~/work/tangram-es/build/perf/bin]% adb shell dumpsys meminfo 10641                                                     [fix-alfons-font-loading]
Applications Memory Usage (kB):
Uptime: 42588929 Realtime: 304383928

** MEMINFO in pid 10641 [com.mapzen.tangram.android] **
                   Pss  Private  Private  Swapped     Heap     Heap     Heap
                 Total    Dirty    Clean    Dirty     Size    Alloc     Free
                ------   ------   ------   ------   ------   ------   ------
  Native Heap    67685    67636        0        0    70388    68814     1573
  Dalvik Heap     5641     5204        0        0    30216    18164    12052
 Dalvik Other      564      564        0        0
        Stack      180      180        0        0
      Gfx dev     5634     5484        0        0
    Other dev        5        0        4        0
     .so mmap     4551      392     3068        0
    .apk mmap       85        0        0        0
    .dex mmap      820        0      816        0
    .oat mmap     1131        0      232        0
    .art mmap      895      580       20        0
   Other mmap       32        4        8        0
   EGL mtrack    47584    47584        0        0
    GL mtrack    11140    11140        0        0
      Unknown      168      168        0        0
        TOTAL   146115   138936     4148        0   100604    86978    13625
```

```
[~/work/tangram-es/build/perf/bin]% adb shell dumpsys meminfo 11340                                                                          [mU]
Applications Memory Usage (kB):
Uptime: 42716223 Realtime: 304711882

** MEMINFO in pid 11340 [com.mapzen.tangram.android] **
                   Pss  Private  Private  Swapped     Heap     Heap     Heap
                 Total    Dirty    Clean    Dirty     Size    Alloc     Free
                ------   ------   ------   ------   ------   ------   ------
  Native Heap    12853    12804        0        0    15620    13862     1757
  Dalvik Heap     5401     4964        0        0    30226    18217    12009
 Dalvik Other      572      572        0        0
        Stack      184      184        0        0
      Gfx dev     5626     5476        0        0
    Other dev        5        0        4        0
     .so mmap     4525      392     3060        0
    .apk mmap       85        0        0        0
    .dex mmap      828        0      824        0
    .oat mmap     1162        0      268        0
    .art mmap     1777      596      664        0
   Other mmap       32        4        8        0
   EGL mtrack    47584    47584        0        0
    GL mtrack    11140    11140        0        0
      Unknown      168      168        0        0
        TOTAL    91942    83884     4828        0    45846    32079    13766

```

**Comment by blair1618** [comment]
Wow that's a lot of memory. Given that our in-memory tile cache is only 32MB, 66MB of font data seems like a number we should try to bring down.

**Comment by hjanetzek** [comment]
@blair1618 ~54MB to be fair. It would be easy to reduce it to a half of that by disposing the buffer held by InputSource once the font is loaded. A bit more work would be necessary to use the same freetype fontface instance for each size, reducing the usage further by 2/3. 
though I think it would be better to use let freetype load fonts via mmap where possible.

**Comment by tallytalwar** [comment]
@hjanetzek in that case this is more of a work outside of tangram and more in alfons.
Are we good to merge this then, and live with ~50MB of extra mem usage for the time being?

**Comment by Songcn** [comment]
Hi, When will these commits be merged into master?  It's really urgent to fix this issue because iOS platform cannot display text now. Thanks a lot.

**Comment by karimnaaji** [comment]
@Songcn , can you let us know exactly how you're using fonts on iOS, are you bundling them with the application you're developing? 
If you bundle them, you should open an issue explaining what errors you're dealing with. Bundled fonts should work in the `master` branch right now.

**Comment by Songcn** [comment]
@karimnaaji , I just cloned master branch to my desktop, then :
1. make make-ios'
2. open build/ios/tangram.xcodeproj
3. Run 'tangram' on my iPad mini2

all steps was done as README.
The app can run but no text labels on the map, Xcode console says:

issing font: error: 1 fonts/NotoSans-Regular.ttf
Missing font: error: 1 fonts/NotoNaskh-Regular.ttf
Missing font: error: 1 fonts/NotoSansHebrew-Regular.ttf
Missing font: error: 1 fonts/DroidSansJapanese.ttf
Missing font: error: 1 fonts/DroidSansFallback.ttf

I guess the fonts are bundled in tangram app by default, right?
 So, was anything wrong during my steps? or any other work I need do? Thanks.

**Comment by Songcn** [comment]
@karimnaaji , By the way OS X version tangram works fine, only iOS version doesn't render labels.

**Comment by karimnaaji** [comment]
Ok will open an issue concerning this.

**Pull request closed without merging**